optuna/storages/in_memory.py

```diff
@@ -187,17 +187,22 @@ def _create_running_trial():
             datetime_complete=None)
 
     def set_trial_state(self, trial_id, state):
-        # type: (int, structs.TrialState) -> None
+        # type: (int, structs.TrialState) -> bool
 
         with self._lock:
             trial = self.trials[trial_id]
             self.check_trial_is_updatable(trial_id, trial.state)
 
+            if state == structs.TrialState.RUNNING and trial.state != structs.TrialState.WAITING:
+                return False
```

Comment:
IMO, we do not need to document the behavior of storage interface for now.

Although the storage interface is public for third party libraries, but it's not public for the end-user of Optuna (because storage object is private in study). And we might need to re-design it for performance improvements.

But this is not a strong opinion. If an another developer think we should, I'll document the behavior.

Reply:
> IMO, we do not need to document the behavior of the storage interface for now. ... And we might re-design it in the future
Thanks for you input, I now agree. It's probably also beyond the scope of this PR.

Reply:
I agree with @c-bata's opinion.

Reply:
I think that the current behavior is tentative. In the future, we would expand this method and become to allow arbitrary state transitions (cf. https://github.com/optuna/optuna/issues/485#issuecomment-522902210 , https://github.com/optuna/optuna/issues/685#issuecomment-554159519)